lib/authenticate_methods.rb

```diff
@@ -3,42 +3,21 @@
 module AuthenticateMethods
 
   def newflow_authenticate_user!
-    # Note to self: remember that SessionsController#start calls `save_new_params_in_session`
     if signed_params.present?
-      if signed_params['role'] == 'student' || account_exists_with_same_email?
-        newflow_use_signed_params
-      else
-        use_signed_params
-      end
-    end
-
-    return if signed_in?
-
-    # Drop signed params from where we will go back after log in so we don't
-    # try to use them again.
-    store_url(url: request_url_without_signed_params)
-
-    if (sp_user = session[:user_from_signed_params]) # student
-      if sp_user['state'] == 'unverified'
-        redirect_to newflow_signup_student_path(request.query_parameters) and return
-      else
-        redirect_to newflow_login_path(request.query_parameters) and return
-      end
-    else # instructor
-      if pre_auth_state && !pre_auth_state.signed_student? && pre_auth_state_email_available?
-        # goes to "old flow"
-        redirect_to main_app.signup_path(request.query_parameters.merge({bpff: true})) and return
-      else
-        redirect_to newflow_login_path(request.query_parameters) and return
-      end
+      use_signed_params_and_redirect
```

Comment:
uggghhh this is a mess of logic.  is every branch of this unit tested?  consider abstracting this logic into a unit testable object w/ callbacks for the redirects.

Reply:
I'm sorry and I agree. But the code isn't always going to be pretty if the requirements or business logic itself is hairy. I do know (from experience 😅) that this is one of the most difficult things about Accounts to wrap your head around as a new Accounts dev. I feel like these two flows (the "signed params" flow and the regular flow) should have been totally separate from the beginning. To do that at this point would be a major undertaking, and I _am_ planning on writing up a card for it, actually.

Reply:
For now, pushed to the next issue and _might_ end up creating a separate issue for it if it turns out to be more greater in scope than what _that_ issue is about: https://github.com/openstax/business-intel/issues/1009#issuecomment-652652720